internal/topology/variables/cluster_variable_validation.go

```diff
@@ -21,87 +21,124 @@ import (
-func validateClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassVariable, validateRequired bool, fldPath *field.Path) field.ErrorList {
+func validateClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassStatusVariable, validateRequired bool, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
 
-	// Build maps for easier and faster access.
-	clusterVariablesMap := getClusterVariablesMap(clusterVariables)
+	// Get a map of ClusterVariables and ensure that variables are not defined more than once in Cluster spec.
+	clusterVariablesMap, err := getClusterVariablesMap(clusterVariables)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, clusterVariables,
+			fmt.Sprintf("cluster variables not valid: %s", err)))
+	}
+	// Get a map of ClusterClassVariables for each variable name and definition.
 	clusterClassVariablesMap := getClusterClassVariablesMap(clusterClassVariables)
 
-	// Validate:
-	// * required variables from the ClusterClass exist on the Cluster.
-	// * all variables in the Cluster are defined in the ClusterClass.
+	// Validate required variables from the ClusterClass exist on the Cluster.
 	if validateRequired {
 		allErrs = append(allErrs, validateRequiredClusterVariables(clusterVariablesMap, clusterClassVariablesMap, fldPath)...)
 	}
-	allErrs = append(allErrs, validateClusterVariablesDefined(clusterVariables, clusterClassVariablesMap, fldPath)...)
-	if len(allErrs) > 0 {
-		return allErrs
-	}
-
-	// Validate all variables from the Cluster.
-	for i := range clusterVariables {
-		clusterVariable := clusterVariables[i]
 
-		// Get schema.
-		// Nb. We already validated above in validateClusterVariablesDefined that there is a
-		// corresponding ClusterClass variable, so we don't have to check it again.
-		clusterClassVariable := clusterClassVariablesMap[clusterVariable.Name]
+	// Validate variables in the Cluster are defined in the ClusterClass and validate variables based on their schema.
+	for i, clusterVariable := range clusterVariables {
+		c := clusterVariable
+		// Check that the variable is defined in the ClusterClass `.status.variables`.
+		clusterClassVariable, found := clusterClassVariablesMap[ccVariableMapKey{clusterVariable.Name, clusterVariable.DefinitionFrom}]
+		if !found {
+			// This will catch cases where a user tries to define a variable without a definitionFrom field. In this
+			// case use a special error.
+			if clusterVariable.DefinitionFrom == emptyVariableDefinitionFrom {
+				allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("name"), clusterVariable.Name,
+					fmt.Sprintf("variable with name %q is not valid. It may not be defined in ClusterClass `status.Variables` or it may require `definitionFrom` in cluster `spec.variables`", clusterVariable.Name))) // TODO: consider if to add ClusterClass name
```

Comment:
Would be nice for users if we can split this error up into two (aka differentiate the cases)